enhancements/machine-api/autoscaling-from-to-zero.md

```diff
@@ -0,0 +1,127 @@
+
+- We want to enable any provider to expose in a machineSet any the details needed by the autoscaler to scale from zero. 
+
+- The autoscaler can then use this details to populate a `schedulernodeinfo.NodeInfo` object enabling scaling from zero. To this end we propose the following annotations for a machineSet:
+
+```
+machine.openshift.io/vCPU
+machine.openshift.io/memoryMb
+machine.openshift.io/GPU
+```
+
+- Details which are non provider specific can be inferred directly from the [machineSet machine template](https://github.com/openshift/machine-api-operator/blob/master/pkg/apis/machine/v1beta1/machineset_types.go#L80) e.g lables, taints.
+
+- The `TemplateNodeInfo` implementation will look very similar to [AWS](https://github.com/openshift/kubernetes-autoscaler/blob/253ee49441750815c70b606f242eb76164d9bdc4/cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go#L321)/[Azure](https://github.com/openshift/kubernetes-autoscaler/blob/253ee49441750815c70b606f242eb76164d9bdc4/cluster-autoscaler/cloudprovider/azure/azure_scale_set.go#L598) autoscaler implementations:
+but  `buildNodeFromTemplate` will get what it needs from the given `nodeGroup`/`machineSet` with zero replicas.
+
+#### Providers implementation
+
+- A new controller is developed in each actuator and run by the machine controller manager.
+
+- It watches machineSets and annotate them.
+
+- Is a per provider choice to decide how to infer the details to be included in the well known annotations.
+
+- For AWS/Azure/GCP the details can be inferred from the instance type set in `machineSet.template.Spec.ProviderSpec`. This can be done dynamically or statically [reusing the autoscaler specific provider implementation](https://github.com/openshift/kubernetes-autoscaler/blob/253ee49441750815c70b606f242eb76164d9bdc4/cluster-autoscaler/cloudprovider/aws/ec2_instance_types.go#L30).
+
+- Providers where there's no instance type semantic (Libvirt, Openstack, baremetal) can decide how best infer their details and leverage the same controller and annotations mechanisim to expose them.
+
+TODO: How to prevent from being scaled to zero a provider which does not support scaling from zero.
+
+## Design Details
+
+### Test Plan
+
+- Unit testing per provider.
+
+- Unit testing for the machine API autoscaler provider.
+
+- Provider agnostic e2e testing coverage in [machine API e2e test suite](https://github.com/openshift/cluster-api-actuator-pkg/tree/a1c4e0f038c06794b7f1436975a7b1b330317c25/pkg/e2e/autoscaler).
+
+
+### Graduation Criteria
+
+### Upgrade / Downgrade Strategy
+
+- This new controller will be included in the `cluster-api-actuator-*` image managed by the CVO and exposed by the machine API Operator.
+
+- When running in the first place by an upgrade it will naturally annotate all the existing machineSets.
+
+### Version Skew Strategy
+
+## Implementation History
+
+## Drawbacks
```

Comment:
Is anybody reliant on the fact that pre an upgrade it would have a min of 1, and could now have 0? (Dunno if this is a real concern but it would be a change in behaviour.)

Reply:
> Is anybody reliant on the fact that pre an upgrade it would have a min of 1, and could now have 0? (Dunno if this is a real concern but it would be a change in behaviour.)

Should you have to opt-in to this new behaviour in some way?

Reply:
I don't think this is necessary. Any existing machineAutoscaler must have a `"machine.openshift.io/cluster-api-autoscaler-node-group-min-size"` set to 1, so this should be honoured by the autoscaler even after support for 0 is introduced.
To opt-in you set your `MachineAutoscalerSpec.MinReplicas=0`